### ISSUE TITLE
Typings bikeshedding

### ISSUE THREAD

**Issue by vendethiel** [gkz/LiveScript#999]
Some time ago (#803) I had started working on a PR to add type hints to the language.

That's still something I miss to this day, so I want to bring it back up.
There were a lot of discussions, so I think we should try to discuss it beforehand.

# Typing symbol

I suggested the symbol `@:` to add a type annotation. Someone suggested `^`. @gkz suggested `::`.

The reason I want(ed) not to use `::` is because it's rewritten to `ID:prototype` when it's parsed directly, at which point we have no info on whether it's going to be part of a signature of just a Parens (we only tag `()` as `PARAM(` and `PARAM)` after seeing `->`).
We'd need to go through the tokens to mark `ID:prototype` as `ASCR` (or whatever else). We'd also need to be careful not to tag in "defaults" – that is, `(x = -> ::) ->`, `(x ? ::) ->`, `(x and ::) ->`, etc...

# where to allow typings

I'd suggest we only allow type annotations in:
 - function declarations (`function` and `->`/`<-`/...)
 - `var`
 - `let`

# generics

I have one suggestion: `[T](xs @: Array[T]) ->`. The reason I suggest `[]` rather than `<>` is mostly for parse-ability (and because I'm used to Scala...). Tagging `[]() ->` is easier than the ambiguities with `<>`, we just have to look for a `]` when re-tagging a `(` as a `PARAM(`, and cycle back to `[` from there (since an array is never callable anyway).
(since `[T]() ->` and `xs[T]() ->` make no sense anyway)

Parsing the `@: Array[T]` part is harder, but still simpler than `@: Array<T>`, where the `>` would be parsed as a BIOP because it's directly followed by a `)`...

--------------

@rhendric I'm interested in your PoV, because it's a sensible topic, considering all the ambiguities to keep in mind :-).

**Comment by vendethiel** [comment]
As is the tradition, I've first implemented this with **~~ LEXER HACKS~~**: 509cf13390336068679c30c4b36f53607bbff849

```
$ ./bin/lsc -m none -bce '(x @: Map[Int, Array[Int]], y @: Int) ->'
(function(x : Map<Int,Array<Int>>, y : Int){});
```

**Comment by rhendric** [comment]
I'm really glad you're still looking into this. I'm also really nervous about adding yet more syntax hacks—I don't know whether LS has already crossed over the line where writing a reasonable parser for it someday is basically impossible, but I'm hoping that nothing we'd do as part of adding types would push it over that line. I want to try to avoid committing any *new* lexical sins, if possible—adding more back-scanning to the `->` strikes me as something to avoid.

The most bikesheddy question: how to spell *has-type*? I guess `@:` is okay. I think I would actually prefer ` :: `, with required spaces on both sides and followed by a token that can start a chain (otherwise, it's still `ID:prototype`). This is not a new lexer sin; other tokens do similar things (` . ` is function composition but `.` is member access, and `++` checks token types on either side).

I also think that we should reuse as much existing LS syntax as possible, where applicable. For that reason, I would propose `Array T` instead of `Array[T]` or `Array<T>`—parse parameterized types as function calls, with the same rules about optional parentheses and comma insertion. This also means there wouldn't be anything hard about parsing `(xs :: Array T) ->` or `(xs :: Array(T)) ->`—it should be handled by existing code.

Given the above three opinions, how to handle generics? If type constructors are applied like functions, I would argue that they should be declared like functions—something like `(T) ::> (xs :: Array T) ->`, where `::>` is a new token for type-level functions. Then `::>` and `->` can share the `PARAM(`-ifying code and no new lexer sins are added here either. (If we go with `@:`, I guess I would advocate `@:>` here, but now we're venturing into real made-up nonsense territory IMO.)

Here are some other things to be worried about (where relevant, this is largely from the perspective of Flow; I'm not as familiar with TypeScript):
* Not having a way to annotate function return types may be limiting, but inserting a return type annotation before the `->` would complicate backtracking. How else could we do that if we decide later on we want to?
* We could get away with writing `{ foo: ?SomeType }` as `foo: SomeType?`, but how would we want to write `{ foo?: SomeType }`?
* `*` is a type, and also already a messily overloaded token in LS.
* Not a worry, more of a don't-forget-to: because the bitwise logic operators in LS are verbose, let's make sure the boolean logic operators can be used instead to construct intersection and union types.

**Comment by vendethiel** [comment]
So, for my current version (still a lexer hack) the return type syntax is `@: Int ->`, with params: `(x @: Int) @: Int ->`.

In the spirit of hacks, to not change Lexer.parameters, if we parse an arrow and our last is an ASCR, we put the token on the side while parsing parameters (there's a good reason I don't want to push it!).

I'm against adding `::>`. It looks bad, and there's no reason we need it. If anything, if we want to go with the way you said (`x :: Array Int` or `x :: Array(Int)`), I think we'd be better off with a 2nd param list, e.g. `(T)(x :: Array T) ->`. I still prefer `[T](x :: Array[T]) ->`, though...

I'm more and more against `::`. For historical reasons, I think we unspace `::`, so `(x :: Int) ->` is `x::Int(->)`.

**Comment by vendethiel** [comment]
> I'm also really nervous about adding yet more syntax hacks

I just wanted to do this as a proof-of-concept.
I mean. It's really hard, because we do so many things at "lexer time" – parameter rewriting, etc. So we need to end up in a "valid" parse states after the lexer ran. Thankfully in this case, it seems we don't end up too poorly (as opposed to what I previously said in my last comment):

```
$ lsc -le '(f :: Map Int (Array Int)) ->'
NEWLINE:\n PARAM(:( ID:f ID:prototype ID:Map ID:Int ( ID:Array ID:Int ) )PARAM:) -> NEWLINE:\n
```

So implementing this syntax is fine. However... For the return type, it's much more complex. even if we use `@:`. We insert `PARAM(`,`)PARAM` when we see a `->`.
My current lexer hack "folds" all the tokens after `@:` into the `@:` (so `(x @: Array[Map Int]) ->` is parsed as `PARAM( ID ASCR )PARAM ->`).

If we want to do it the right way, that means we need to change `Lexer.parameters` to try and go back "N" tokens (an indefinite number of them) to try and find `) @:` so that we know we're parsing a signature...

On the other hand, if we go for a syntax like `@: Map[Int, Array[String]] ->`, the parsing is much easier, we either have a single lookbehind (a simple id) or until we find a matching `[` for that `]` (`Array[Int]`). We also have to throw away some dots (since `x[3]` is `ID DOT [ STRNUM ]`). At least until unions etc enter the field...

> I don't know whether LS has already crossed over the line where writing a reasonable parser for it someday is basically impossible, but I'm hoping that nothing we'd do as part of adding types would push it over that line. I want to try to avoid committing any new lexical sins, if possible—adding more back-scanning to the -> strikes me as something to avoid.

Well, I don't think it would be impossible. We'd just need a parser that'd be less confused as to what it's looking at.

> This also means there wouldn't be anything hard about parsing (xs :: Array T) -> or (xs :: Array(T)) ->—it should be handled by existing code.

As demonstrated, it doesn't, because some rules don't apply between `PARAM(` and `)PARAM`.

> Not having a way to annotate function return types may be limiting, but inserting a return type annotation before the -> would complicate backtracking. How else could we do that if we decide later on we want to?

See ^

> We could get away with writing { foo: ?SomeType } as foo: SomeType?, but how would we want to write { foo?: SomeType }?

I have no idea how these two are different. Optional field vs nullable type?

> * is a type, and also already a messily overloaded token in LS.

It's a type??

**Comment by ozra** [comment]
I like the idea of spaced `::` — in my pet language there are hoards of situation where punctuation differ in meaning, depending on how it's spaced. Surprisingly I haven't found it confusing at all. YMMV.
Spaced `::` means "here comes a type". And that's that.

```livescript
# Obvious. Typed variable
x :: Int = 0

# Type aliases!?
MyBarCombo = :: BoozeBar<Qwo, Bzt> | (Barable<Qwo> & Boozable<Bzt>)

# Of course, param types and return type
foo = (a :: String, b :: Any?, c :: MyBarCombo) :: SuperBool ->
    console.log "#{a}, #{b ? ""}, #{c}"
    super-true

# casting / coercion!?
console.log "A real number:", ((47 * 23) :: Real)

# Extending on the alias above, crazy:
MyClass = ::
    (foo) ->
        @foo = foo + "rules"
        @i-give-up = "now"
```

**Comment by determin1st** [comment]
didnt dive into all the comments, sorry, but why not add some kind of parse logic marker at the top of the file with code, like

"use strict, ts_compat"

so it will be converted to globally accessible flag and checked where the compile logic works? so the livescript could be extended to anything.. or maybe it's an utopia)

**Comment by vendethiel** [comment]
> in my pet language there are hoards of situation where punctuation differ in meaning, depending on how it's spaced.

in LS, we have:
 - calls use space (`f a b`), except for literals using ACI.
 - existential operator: `a? b` vs `a ? b`.
 - lines: `a b .c`, etc.
 - dot: `a.b` vs `a . b`
 - regex (`f /a/` vs `f / a /` => disallowed)
 - `++` (`f a++ b` => `f(a++, b)` vs `f a ++ b` => `f(a.concat(b))`)
 - keywords (`require ! a` vs `require! a`).
 - ADI/ACI (`a [b]` vs `a[b]`)
 - biop

So that ship sailed a loooooooooooooong time ago. Doesn't mean we **need** anymore of them. But yes, we *can* have dually-spaced `::` mean "type ascription".

```livescript
# Obvious. Typed variable
x :: Int = 0
# vendethiel: my goal is to only allow this in let, var, and const.

# Type aliases!?
MyBarCombo = :: BoozeBar<Qwo, Bzt> | (Barable<Qwo> & Boozable<Bzt>)
# vendethiel: probably, but with a keyword. Also, no <>.

# Of course, param types and return type
foo = (a :: String, b :: Any?, c :: MyBarCombo) :: SuperBool ->
    console.log "#{a}, #{b ? ""}, #{c}"
    super-true
# vendethiel: that works locally.

# casting / coercion!?
console.log "A real number:", ((47 * 23) :: Real)
# vendethiel: maybe? it currently parses but doesn't do anything.

# Extending on the alias above, crazy:
MyClass = ::
    (foo) ->
        @foo = foo + "rules"
        @i-give-up = "now"
# vendethiel: what's that supposed to be?
```

**Comment by wryk** [comment]
## Type constructor
I prefer application syntax `Map Int, Array(Int)` over `Map[Int, Array[Int]]` but livescript application syntax use commas :
```ls
# only one parameter, it's fine
(x :: Map Int, Int) ->

# multiple parameters, commas confusion :/
(x :: Map Int, Int, y :: Int) ->

# we need use parens to fix this ...
(x :: (Map Int, Int), y :: Int) ->
# or
(x :: Map(Int, Int), y :: Int) ->

# with haskell-like application syntax it's better but it's out of scope for this issue
(x :: Map Int Int, y :: Int) ->
```

## Type ascription
I prefer `::` over `@:` too. But adding more space-based operators isn't that great ...
We should remove `::` as a shortcut for `prototype`. (I'm only half ironic)

## Return type
Proposed return type `() :: Int ->` looks fine.

## Type parameters
I prefer the postcircumfix brackets `[A] (x :: A) ->` over the postcircumfix parens with a custom separator `(A) ::> (x :: A) ->`.
```

## Higher-order function
I didn't see any example with them :/
```ls
# something like that ?
(f :: ((x :: Int) :: String), x :: Int) :: String ->
```

**Comment by vendethiel** [comment]
of importance: we could [probably] have `(T, F)(a :: T, a :: f) ->` if we wanted it.

**Comment by rhendric** [comment]
### Type params

I prefer `[A](x :: A) ->` over `(A)(x :: A) ->`; there's less chance for ambiguity with the former. The reason I think a separating operator is a good idea is because right now, the lexer reaches backwards from the `->` all the way to the opening `(` of the parameter list, and that's bad enough. I don't want to make that operation even less local by making it extend backwards further to capture another set of brackets of any type. If `[A](x :: A) ->` can be done only by changing how `]` is lexed—say, when it is immediately followed by a no-space `(`—then I'm more okay with it.

What about something like `forall A, (x :: A) ->`? A feature this big might be worth a keyword or two. That also makes generic types that don't involve functions possible: `forall A, [A, A]`, for example, the type of homogeneous pairs. And this is very left-to-right-parsing friendly.

### HOF and return types

I don't like the double `::` for higher-order functions—much better if we can use an arrow the way I think both Flow and TypeScript do. But then this leads to precedence problems if we are annotating return types the way it has been previously proposed:
```ls
map = forall (A, B), (f :: A -> B, arr :: Array A) :: Array B -> ...
# or
map = [A, B](f :: A -> B, arr :: Array A) :: Array B -> ...
```
If `::` has higher precedence than `->`, then `f :: A -> B` won't work; but if it has lower precedence, then the return type of the function will look like `Array B -> ...` instead of `Array B`.

I already had issues with that notation for return types because, again, it extends the non-locality of the `->` lexer rewriting—the lexer has to reverse-parse a type, but only if that type is preceded by a `::`, all before reverse-parsing a parameter list—ugly! I think we need some other alternatives for return types, perhaps following the `->` instead of preceding it. Not sure what that should be though.

### Flow stuff

Yes, in Flow, `{ foo: ?SomeType }` is a field with a nullable (or undefined-able) type, and `{ foo?: SomeType }` is an optional field. Flow cares about the difference because it changes whether `foo in obj` is a discriminator for determining if `obj` is of that type.

`*` is what Flow calls the ‘existential type’. It's a fill-in-the-blank request for the type checker—i.e., *something* (that isn't `any` or `mixed`) should go here, you figure out what.

### @determin1st's suggestion

Supporting multiple syntaxes, however you do it (through in-source pragmas like what you suggest, configuration files, or compiler flags), exponentially increases what the language maintainers need to support. LiveScript's one syntax is barely being maintained right now; I am strongly against dividing our efforts further unless there is a renaissance in people contributing to the code.

### Classes

We should also think about these. LS classes don't create ES6 classes, so they should be typeable as `(constructor types) -> { all declared members in the class }`. I guess, for starters, that could be done manually with a `var` declaration above the `class` declaration, but that's pretty bad long-term due to the redundancy. But if we want to do this the right way, we also need syntax for annotating the types of properties in objects, since that's how LS classes are declared. `member :: Type: value` is shorthand for `member :: { Type: value }`, so that syntax is out. We could maybe special-case `(member :: Type): value`, if we're sure we'll never support `something :: Type` as an expression? Or we need some other idea.

**Comment by vendethiel** [comment]
# Type params

> What about something like forall A, (x :: A) ->? A feature this big might be worth a keyword or two. That also makes generic types that don't involve functions possible: forall A, [A, A], for example, the type of homogeneous pairs.

LS tends to follow Coco in the "Fewer keywords" mantra.
It's a bit weird because we're really ditching the likes of C++, Scala, TypeScript, etc, while going for a Haskell-like approach, where they separate the type signature and the body.

> And this is very left-to-right-parsing friendly.

Well, `forall (A, B) ->` currently parses relatively cleanly (`ID:forall ( ID , ID )`, though it gets weirder after rewriting...), so it might be doable. I still prefer getting a syntax that resembles Flow/TS that we are targeting, though.


# HOF and return types

> I don't like the double :: for higher-order functions—much better if we can use an arrow the way I think both Flow and TypeScript do (and again, let's reuse things that already exist in LiveScript). But then this leads to precedence problems if we are annotating return types the way it has been previously proposed:

If we do that, we can just able forget the idea of doing this properly in the parser and not in the lexer, methinks... Also then it becomes weird that the function's own return type does not use that keyword (though it's the same as in TS, e.g. `function(f: int => int): int`).

> I think we need some other alternatives for return types, perhaps following the -> instead of preceding it.

I don't think so:
 - Make it look weird considering our current implicit block (`-> 5` vs `-> :: Int => 5` – or something)
 - Make the lexer/parser have to gobble N tokens, probably until a newline (`-> Int & Map Int, String`).

# Flow stuff


> Yes, in Flow, { foo: ?SomeType } is a field with a nullable (or undefined-able) type, and { foo?: SomeType } is an optional field. Flow cares about the difference because it changes whether foo in obj is a discriminator for determining if obj is of that type.

I see. There's no such difference in TS AFAIK. I find it really ugly, but hey. We havn't even started to bikeshed on how to type heterogeneous objects yet.

> * is what Flow calls the ‘existential type’. It's a fill-in-the-blank request for the type checker—i.e., something (that isn't any or mixed) should go here, you figure out what.

Ok. Thankfully, we already have a placeholder thingie, which is what the ML family also uses: `_`.

# Classes

See the comment on heterogeneous objects. They fall under the same category I'd say.

> We could maybe special-case (member :: Type): value

Note that this is not enough.

```livescript
class C
  (a :: Array Int): []
  -> @a.push 3
console.log C.new.a # prints [3]
console.log C.new.a # prints [3, 3]
```

**Comment by rhendric** [comment]
> Note that this is not enough.

I don't understand that last example at all. More words, please?

**Comment by vendethiel** [comment]
We need a way to type instance variables, not only prototype variables.

Are you available on IRC for a few minutes?

**Comment by rhendric** [comment]
```
[14:50] == rhendric [18bcc478@gateway/web/freenode/ip.24.188.196.120] has joined #livescript
[14:51] <Ven``> rhendric: o/
[14:51] <rhendric> Hi!
[14:51] <Ven``> I was hoping you could join IRC, because discussing this on github is proving a bit complex
[14:52] <rhendric> Understandable, but I hope that most of this conversation can be both public and in one place, so if we have to have IRC chats I think we should post the logs in the issue at least.
[14:52] <Ven``> oh, I definitely want to paste logs afterwards!
[14:52] <rhendric> Excellent.
[14:53] <Ven``> ok, so first, I really wasn't expecting it'd come to something like this when I started that PR in late 2015.
[14:53] <Ven``> Though to be fair, my PR then was just a sribble put together for demonstration's sake
[14:54] * Ven`` has taken a long time to come to terms with gkz's non-{interest,involvment} with the project
[14:54] <Ven``> ah, well, it's to be expected that people move on after >5 years.
[14:54] <rhendric> Especially given the state of the ecosystem.
[14:54] <rhendric> LiveScript is getting a little crowded out of its niche.
[14:55] <Ven``> You think so? I think it's less crowded than it used to.
[14:55] <Ven``> there's not a new alt-js every other day :-).
[14:55] <rhendric> This is a separate conversation (that I'm happy to have, but don't want to inadvertantly digress).
[14:55] <rhendric> But yeah, I'm sympathetic to gkz's position at least.
[14:55] <Ven``> Alright, fair enough.
[14:56] <Ven``> we can discuss that later or at some other point :-).
[14:56] * rhendric nods
[14:56] == Ven` [~textual@118.59.28.93.rev.sfr.net] has joined #livescript
[14:57] <Ven``> we have (potentially) a lot to discuss. Should we start by bikeshedding, discussing viable implementations, ..?
[14:59] <rhendric> Not sure those conversations are separable. Implementation should inform design at least somewhat. Maybe we should start with goals?
[14:59] <Ven``> OK, fair enough.
[14:59] <rhendric> As in, is this the beginning of an eventual full support of types in Flow and/or TypeScript format?
[14:59] <rhendric> Is this a stopgap measure to make things a little more usable?
[15:00] <rhendric> Something else or in between?
[15:00] <Ven``> Definitely full support. Though not necessarily in the very beginning.
[15:00] <Ven`> I don't mind if I don't get e.g. object types and class types in the early beginning.
[15:01] <Ven`> Inference itself would help alleviate that
[15:01] <rhendric> Ideally, yes.
[15:02] <Ven`> For now, I want to get types on function args/function returns/let/var/const.
[15:02] <Ven`> simple types and generics.
[15:02] <rhendric> But we should also keep an eye on the rest of what we eventually want to support, in that case.
[15:02] <Ven`> Oh, definitely.
[15:02] <Ven`> it's tough. the syntax baggage of LS is... quite heavy
[15:03] <rhendric> Yes, that.
[15:03] <Ven`> I'd be interested if we could come up with a way to type objects that's not hell to parse.
[15:03] <rhendric> So.
[15:03] <Ven`> (and doesn't require to do it all in a @do-ascr in the lexer...)
[15:03] <rhendric> A lot of these things become much easier if we go full Haskell and separate types from values entirely.
[15:04] <Ven`> We definitely should, but I don't see much we need to do to get that..?
[15:04] <rhendric> There are some usability issues with that, clearly, but if we eventually want to support the moving target that is ‘anything that TypeScript/Flow support’, I think that's our best bet.
[15:05] <rhendric> Not much we need to do, agreed.
[15:05] <Ven`> What are the issues you're thinking of?
[15:05] <rhendric> Well, not as much.
[15:05] <rhendric> When I say separate, I mean no `(a :: A) :: B => a-to-b a`.
[15:06] <rhendric> Instead `f :: (A) -> B`
[15:06] <rhendric> `f = (a) -> a-to-b a
[15:06] <Ven`> OK. So really full Haskell.
[15:06] <rhendric> Yeah.
[15:06] <Ven`> it makes some stuff neater. It makes some stuff harder.
[15:06] <Ven`> How do you spell a let with types?
[15:06] <Ven`> How do you spell a typed instance variable?
[15:06] <Ven`> How do you spell a ES6-class-like method?
[15:08] <rhendric> let: not sure if necessary?
[15:08] <rhendric> But if necessary, ascription inside the let block, I guess.
[15:09] <rhendric> Maybe TypeScript isn't as good as Flow is in letting stuff like that be implicitly typed.
[15:09] <rhendric> typed instance variable: inside an LS class, just add a declaration.
[15:09] <Ven`> I think it's not as good, but I'd rather target TS than Flow, TBH.
[15:09] <Ven`> It seems like the community picked a side...
[15:10] <Ven`> (Windows support played a huge part there IMHO)
[15:11] <rhendric> I think in communities that use React at least, Flow is still dominant.
[15:11] <rhendric> Possibly others, although it's hard to see past the bias of my own experience.
[15:11] <Ven`> Same for my echo chamber :-).
[15:12] == Ven` [~textual@118.59.28.93.rev.sfr.net] has quit [Read error: Connection reset by peer]
[15:12] == Ven` [~textual@118.59.28.93.rev.sfr.net] has joined #livescript
[15:12] <rhendric> Anyway, you think TypeScript would require a function declared and used only once to have explicit types on its parameters to be typed inside?
[15:14] <Ven`> https://imgur.com/a/felN9
[15:15] <Ven`> yup.
[15:15] <rhendric> Gross.
[15:15] <Ven`> they probably do that for "backwards compat reasons"
[15:15] <rhendric> Probably.
[15:16] <rhendric> It makes sense given the gestalt I have of TS philosophy, but still.
[15:16] <rhendric> It makes our work really difficult, because what if there are other instances of LS constructing code that TS would need to have annotated?
[15:17] <Ven`> I think it does objects correctly.
[15:17] <Ven`> though we need a way to write union&intersection types. that goes for both flow & TS
[15:17] <Ven`> (I think we both agree Flow is superior to TS, it's just about the size of the community
[15:17] <Ven`> )
[15:17] <rhendric> What about LS's util functions?
[15:17] <rhendric> Will things like in$ need to be annotated?
[15:18] <Ven`> Tough call. Probably? This is where flow comments would prove useful, I suppose...
[15:18] * Ven` doesn't want to introduce yet another compiler flag like --types
[15:18] * rhendric doesn't either
[15:19] <rhendric> Times like this, that babel backend would really come in handy.
[15:19] <Ven`> Yup, I was really thinking the same :-P.
[15:20] <Ven`> Having a sane AST and a sane codegen would help as well
[15:20] * Ven` mourns CSR for a bit
[15:20] <Ven`> I havn't looked at livescript-next too much, actually. Anyway, that's enough disgression.
[15:21] <Ven`> so, I'll do some digging to see if TS is able to infer anything at all.
[15:21] <Ven`> https://www.typescriptlang.org/docs/handbook/type-inference.html seems like it does bottom-up type derivation
[15:21] <Ven`> but no further "type inference"
[15:21] <Ven`> That's an issue with your proposal, I think, I don't really want to have to name all the functions I'm going to type.
[15:22] <Ven`> I mean, an issue with the Haskell-style declare-then-define
[15:22] <rhendric> Yeah. Like I said, usability issues.
[15:22] <rhendric> I know it's less powerful than the alternative.
[15:23] <rhendric> I'm just not sure the more powerful thing can be made compatible with LiveScript's syntax without creating a messy headache.
[15:24] <Ven`> "more powerful thing" being?
[15:24] <rhendric> ‘Mixfix’ type annotations.
[15:25] <Ven`> you mean infix :: or @: ?
[15:25] <rhendric> Yeah.
[15:25] <rhendric> Along with whatever we do for return types for functions.
[15:29] <Ven`> So you'd propose to still do it the Haskell way and all?
[15:32] <Ven`> I don't think that way would make the feature that useful to me.
[15:32] <Ven`> ....unless we go totally crazy and re-introduce `where` and allow type annotations there :P. Full haskell indeed.
[15:32] <rhendric> Ha, yeah, that seems off the table.
[15:32] * Ven` wouldn't be against reintroducing where, but that's another issue altogether!
[15:32] <rhendric> If it's not useful, then no point in trying to do it that way.
[15:33] <Ven`> well, the haskell way would mean that for e.g. every .map I want to write needs a declared function.
[15:33] <Ven`> That's a verbosity LS has not used me to :P.
[15:34] <rhendric> The prospect of `complicated LS expression with (complicated destructuring :: complicated type) :: more complicated type -> function body` freaks me out a little.
[15:34] <Ven`> no, no, I think that's a non-goal.
[15:34] <rhendric> Yeah, I was operating on the assumption that little map one-liners could be inferred by the type checker, but I guess that won't work for TS.
[15:35] <Ven`> You don't write `{a :: number} = x`. That's one case where both TS and Flow do the right thing.
[15:35] <Ven`> https://imgur.com/a/ArmSM
[15:35] <Ven`> (it doesn't show b's type because it's not selected)
[15:36] <rhendric> So: let's annotate `objs.map(({ a, b }) -> (x) -> a*x + b)`
[15:36] <Ven`> actually this is a better example: https://imgur.com/a/KI0wI
[15:37] <Ven`> Proposed:
[15:38] <Ven`> https://gist.github.com/vendethiel/49ff86793d796c535e45de75ddc58f62
[15:40] <rhendric> For that matter, maybe:
[15:41] <Ven`> (we don't do anything with `type`, we just translate it to an alias. Hopefully those look the same for TS and Flow....)
[15:41] <rhendric> type f = Function { a :: number, b :: number }, Function number, number
[15:41] <rhendric> objs.map(({ a, b }) -> (x) -> a*x + b :: f)
[15:42] <rhendric> And *then* we can replace Function with -> without conflict :-).
[15:42] <Ven`> uhhh... parsing f there to be the whole function's type is gonna prove **difficult**
[15:42] <rhendric> Is it?
[15:42] <rhendric> Just give :: very low precedence?
[15:42] <Ven`> precedence isn't gonna matter, we added parentheses there.
[15:43] <Ven`> (also my progress on :: was further stalled because <<< might be somehwere rewritten to ::, I didn't check again)
[15:43] <rhendric> (Whatever, mentally replacing :: with @: if necessary isn't a big deal I don't think.)
[15:43] <Ven`> (I just wanted to make a comment about it)
[15:43] <rhendric> I don't understand what the problem is still.
[15:44] <Ven`> The problem for a :: there?
[15:44] <rhendric> I guess? Is that the problem?
[15:45] <Ven`> first off, it potentially conflicts with casts. Second, it's not clear which function it applies to. Third: we rewrite `(x) -> (y) -> x + y :: f` to `(x) -> ((y) -> (x + y.prototype(f)))` before the parser ever sees it.
[15:45] <Ven`> So there's nothing about "precedence" etc to be applied.
[15:47] <rhendric> That rewrite can be made to stop at ASCR tokens, can't it?
[15:47] <rhendric> What's the cast conflict?
[15:47] <Ven`> also I don't see how we'd compile your code, outside of a cast?
[15:47] <Ven`> the cast conflict: well, we might want a + b :: Int to mean `(int)(a + b)`
[15:48] <Ven`> The rewrite can be made to stop at ASCR tokens, but then how do you type only the inner function?
[15:48] <rhendric> Oh, does TS allow that?
[15:48] <Ven`> I think so. At least there's a "x as Type", swift-like
[15:48] <Ven`> (or c# like rather)
[15:48] <rhendric> With Flow, you can only (upcast? downcast? Whichever one is the safe one) that way.
[15:48] <Ven`> it depends on the type's variance :-)
[15:49] <rhendric> You type only the inner function by adding parentheses, if you want.
[15:50] <Ven`> it still seems to me like we lose reading comprehension there. By putting the important infos on the wayside
[15:50] <rhendric> A little, sure.
[15:51] <rhendric> But your way, the arrow rewriter in the lexer has to jump over the Function Int Int before it gets to the PARAM( rewriting it needs to do.
[15:51] <Ven`> that's not my way, that's your way ;-0
[15:51] <Ven`> my way is with [].
[15:51] <rhendric> Ooookay, sure.
[15:52] <rhendric> So if we go full Scala, we now have two different ‘expression’ languages in LiveScript.
[15:52] <rhendric> And how do you notate tuples? A `Tuple` special type constructor like `Function`?
[15:53] <Ven`> TS doesn't have tuples
[15:53] <Ven`> oh, they do actually
[15:53] <Ven`> they write [a, b], huh. How is it in Flow?
[15:54] <rhendric> Same.
[15:54] <Ven`> I'm surprised they didn't go for the "obvious" (a, b)
[15:54] <Ven`> Okay.
[15:55] <rhendric> I have to go in a few minutes; is there a nice place to conclude this?
[15:56] <Ven`> Well, from reading this, apparently not
[15:56] <rhendric> I think we've agreed that both TS and Flow should be targets; that we want to support everything possible eventually; and that a pure Haskell-like approach isn't viable (but maybe something that is almost that approach with some very simple infix :: support is?).
[15:56] <rhendric> Other than that, a lot of unknowns and maybes.
[15:57] <Ven`> https://gist.github.com/vendethiel/ae15873a428627de640a32cdefe83ce2 is really not that bad.
[15:57] <Ven`> We could have :: (Int -> Int), if we force parentheses to be used.
[15:57] <Ven`> I think that's totally viable
[15:58] <Ven`> ... but we need to make -> lookbehind more complex, yet again.
[15:58] * Ven` didn't realize it had already been an hour...
[15:58] <rhendric> I personally am willing to give up a fair amount of visual appeal in order to have the lexer not get more complex.
[15:59] <Ven`> actually no, I'm totally wrong.
[15:59] <Ven`> @parameters doesn't need to be made any more complex.
[15:59] <Ven`> We also have the option of ASCR doing some parsing. Like a sub-language, much like literals or regexes.
[16:00] <rhendric> I have to go.
[16:00] <rhendric> I'll happily review any further POCs you want to invest in.
[16:00] <Ven`> I suppose on the issue tracker
[16:00] <rhendric> You've heard the objections I anticipate making.
[16:00] <Ven`> aye.
[16:01] <rhendric> Okay, good hunting!
```

**Comment by rhendric** [comment]
I wrote my own proof of concept, with tests (I've checked the resulting code on typescriptlang.org; looks good as an initial stab). See [68e836267](/rhendric/LiveScript/commit/68e836267).

It's very hacky also, but all the hacks are in ast.ls instead of lexer.ls. Notable bikeshed decisions made here:
* Spaced `::` for *has-type*
* Two ways to annotate function types: as a single large unit by annotating the entire function expression, or by annotating the individual parameters and the last line of the function body (somewhat unconventional, I know, but works well for one-liners)
* Type parameters declared using a `forall` ‘type’ which gets special handling in ast.js (it lexes just like a regular identifier, and so doesn't need to be a reserved keyword elsewhere in LS code)

**Comment by vendethiel** [comment]
Nicely done!

**Comment by vendethiel** [comment]
I'm gonna go full crazy for a bit... [@rhendric first proposed Haskell-style annotations:](https://gist.github.com/vendethiel/5b830f94b5cf838e66c61ea629e31d0c#file-irclog-livescript-2018-01-18-txt-L27-L28)

```livescript
map :: forall A, B, (Array(A), (A) -> B) -> Array B
map = (as, f) -> as.map(f)

mapper :: Int -> Int -> Int 
mapper = (x) -> (y) -> x + y
map [] mapper
```

I think this is impractical as it means you need to declare a new variable for every function.

I want to mention a way this **could** be solved using a Haskell-style where (...which we add a few years back), if made a bit smarter:

```livescript
map [] mapper
  where mapper :: Int -> Int -> Int
             mapper = (x) -> (y) -> x + y
```

would not bother me, if we made `where` recognize variables that are mentioned only once, and did literal replacements.

**Comment by ozra** [comment]
@vendethiel, I should have been more clear as to my intention with the examples. They were more of what would work without clash with the simple definition "after ` :: ` comes type expression". They were actually _not_ meant as actual desired syntax. _Sorry about the confusion._

>  vendethiel: my goal is to only allow this in let, var, and const.

[var decl] - That sounds preferable to me too.

>  vendethiel: probably, but with a keyword. Also, no <>.

[type alias] - I fully agree here too, except `<>` or `[]` are fine by me.

>  vendethiel: maybe? it currently parses but doesn't do anything.

["casting"] - I figure it can't harm to allow the notation should some one find a good use for it for a static analysis tool or such. If TypeScript is considered _the_ specific target for LS type notation (_which I would find perfectly reasonable_), then of course  any notation that won't forseeably be introduced, or already exist, in TS would maybe be confusing.

>  vendethiel: what's that supposed to be?

[class declaration] - Based on the already crazy example of type alias, this would simply signify a class declaration. But again, not wanted in practice.

I just prefer the idea proposed of spaced `::` over `@:` etc. variations.

Regarding generics, I _much_ prefer the `Foo[Bar, Zoo]` over `Foo(Bar, Zoo)`. Much clearer separation of distinct concepts. I could go into a lengthy motivation here concerning the human language processing unit and the need for redundant information (in contrast to my earlier examples), but... when someone's prepared to put in the time to implement type annotations, I'm happy regardless of the syntax in the end.

**Comment by vendethiel** [comment]
> [type alias] - I fully agree here too, except <> or [] are fine by me.

I've always prefered matching brackets, but fine.



> then of course any notation that won't forseeably be introduced, or already exist, in TS would maybe be confusing.

we could have it mean TS' `varname as Type`.

**Comment by vendethiel** [comment]
https://github.com/Microsoft/TypeScript/pull/21316
https://github.com/Microsoft/TypeScript/pull/21496

TypeScript seems to be moving... in many directions at once.

How the hell are we writing `type ReturnType<T> = T extends (...args: any[]) => infer R ? R : T;` in LS, except for

```livescript
`type ReturnType<T> = T extends (...args: any[]) => infer R ? R : T;`
```

**Comment by rhendric** [comment]
I'm not at all sure this is a good idea, but continuing in the vein of my proof-of-concept, the below comes remarkably close to compiling:
```ls
type ReturnType = forall T, if T extends ((...Array any) -> infer R) then R else T
```

**Comment by vendethiel** [comment]
> I'm not at all sure this is a good idea, but continuing in the vein of my proof-of-concept, the below comes remarkably close to compiling:

If anything, I think I'd prefer it if we had a `Type`-like class that'd be basically a pattern matcher where your AST hack were extracted to, that'd recursively rewrite calls and al to types.

**Comment by rhendric** [comment]
That would be a clean way to encapsulate things. However, I think I'd want to avoid a proliferation of `TypeFun`, `TypeObj`, etc. classes, duplicating the existing AST class family. And I want to reiterate that existing shorthand notations (`foo: string` == `{ foo: string }`, etc.) should work as well in type expressions as they do in value expressions, and a lot of that code lives in the existing AST classes. So that's a thing to work around.

**Comment by vendethiel** [comment]
> However, I think I'd want to avoid a proliferation of TypeFun, TypeObj, etc. classes, duplicating the existing AST class family.

That's not what I'd envisioned. The Type::compileNode would just generate JS from its content, recursively descending into its AST fragment.

**Comment by rhendric** [comment]
Could work! That approach might necessitate pulling some logic that lives in existing `compileNode`s out into their own methods, but that's not a bad thing.

**Comment by vendethiel** [comment]
> Could work! That approach might necessitate pulling some logic that lives in existing compileNodes out into their own methods, but that's not a bad thing.

Indeed, I'd argue that's a good thing, and untangling type/real calls code is the only sane way forward imho.

**Comment by vendethiel** [comment]
@rhendric would you consider either 1) allowing type declaration beforehand, like my `where` example and/or 2) being able to declare the return type as `-> :: x`, which works better for non-single-line expressions (imho – because you can read the whole signature in one go)

**Comment by rhendric** [comment]
I'm open to both of those possibilities, depending on how they're executed. I do like the idea of (2) over `(args) :: ReturnType ->`. Might even be nice in one-liner form; something like `(it :: any) -> :: String; it.to-string!`. I'm neutral on `where`, but could see myself loving it (or hating it) after more consideration.

**Comment by vendethiel** [comment]
> Might even be nice in one-liner form; something like `(it :: any) -> :: string; it.to-string!`

Ah, but... obviously, as it currently closes implicit functions, it's non-viable. I don't think we can keep both syntaxes. `-> :: Number` is ambiguous (is number the return type, or the whole function type)

**Comment by rhendric** [comment]
Why is that non-viable? We'd just change the lexer to not close implicit functions at `::` and add a grammar rule to `Block` to support type ascriptions appearing as the first line of the block. Then `-> :: T` would make `T` the return type, and `(->) :: T` would make `T` the whole function type.

**Comment by vendethiel** [comment]
> Why is that non-viable?

Okay, not non-viable, sorry – I'm just biased against that syntax.

**Comment by rhendric** [comment]
Aw, really? I'm starting to like it better than my original proposals. Having support for all of

```ls
fun-with-types = (a :: A, b :: B) -> :: C; ...
fun-with-types = (a :: A, b :: B) -> :: C
  ...
fun-with-types = (a :: A, b :: B) ->
  :: C
  ...
```
feels nice and uniform to me, and all of these seem more readable than
```ls
fun-with-types = (a, b) -> ... :: (A, B) -> C
fun-with-types = (a :: A, b :: B) -> (... :: C)
fun-with-types = (a :: A, b :: B) ->
  ... :: C
```
which were the hacks I first tried out.

**Comment by vendethiel** [comment]
> and add a grammar production to Block to support type ascriptions appearing as the first line of the block.

Bah, I'm really starting to consider adding an `Ascr` node in ast.ls to support that. (`if block.lines.0 instanceof Ascr then my-return-type = block.lines.shift!`. It'd probably also be the place to extract the `if o.in-type-expr` code.

**Comment by rhendric** [comment]
I pushed [a commit](/rhendric/LiveScript/commit/22e859e4d) to [my proof of concept branch](/rhendric/LiveScript/tree/types) supporting `-> ::`-style function annotations. I kept support for the more awkward alternatives, because barring another syntax <s>abuse</s> innovation, `forall` is still needed to declare type parameters.

**Comment by vendethiel** [comment]
> because barring another syntax abuse innovation, forall is still needed to declare type parameters.

Is this not acceptable for some reason? Obviously a breaking change, but...

```livescript
map = forall A, B, (f :: (A) -> B, arr :: Array A) -> :: Array B
  arr.map f
```

**Comment by rhendric** [comment]
> Is this not acceptable for some reason?

Technically, that could probably be made to work. Aesthetically, I'm biased against it because of how freely it mixes type and value symbols without clear syntactic dividers between them (`::`). Instead of knowing that `forall` is a special type constructor, but basically the same sort of thing as `Array` or `Map` in type expressions, you would now have to know that `forall` is a special language keyword that introduces type variables into the scope of a child value expression, which is quite its own beast.

**Comment by vendethiel** [comment]
> you would now have to know that forall is a special language keyword that introduces type variables into the scope of a child value expression

That ship has sailed already, though, no? even when writing `-> o :: forall A, B, ...`, `forall` is special because any other identifiers would result in a parameterized type

**Comment by rhendric** [comment]
Maybe I'm splitting hairs, but to me the difference seems large: `forall` as a type constructor is special, yes, but special *as a type constructor*. To a reader, it behaves syntactically like a type constructor—valid in a type expression, and its children are type expressions. It's only special semantically. `forall` as a keyword in a value expression is a unique construction. You can't think of it syntactically as a function, because most of its arguments are types; and it isn't syntactically a type constructor because it overall is a value, and so is its last argument. It'd be the first special form that works with types, besides `::`. I'm hoping to avoid introducing such forms, because (given my preference for reusing value expression syntax for type expressions) I think it's going to be important to be very clear when you're looking at a type and when you're looking at a value. If `::` is the universal this-is-a-type indicator, great. If we add more things like keyword `forall` that can indicate types, I think that's a big step towards anarchy and confusion.

**Comment by rhendric** [comment]
I should also add that I expect there to be more special type constructors; see, for example, `infer` from several comments up. (That one is only special in that it would compile to `infer T` instead of `infer<T>`, but still.)

**Comment by vendethiel** [comment]
> I should also add that I expect there to be more special type constructors; see, for example, infer from several comments up. (That one is only special in that it would compile to infer T instead of infer<T>, but still.)

Fair enough, but I don't see any other that need to be an introducer. Maybe if Flow or TypeScript start supporting existential qualifications, but...
I see your point though. We're making a construct that "starts outside" the type part, and wrap it back into it. That's a precedent... But clearly I don't want a different syntax to introduce type parameters.

There's no real "good" place, except if we want to... pluck them out as we're reading it (basically like TS's `infer` works/will work): rewrite `(f :: (forall A) -> forall B, xs :: Array B) -> :: Array B`. That way we never mix type-level and value-level, but I really don't like the way it looks, it's a bit awkward, and it makes reading the type signature harder (imho).

---------------


I'd appreciate if a few more people voiced their opinions, or threw in some other ideas :-).

**Comment by rhendric** [comment]
(Starting to wonder if perhaps `[A, B](a :: A, b :: B) -> ...` is the least of all evils after all...)

**Comment by vendethiel** [comment]
Surely you mean the least `forall` evils.. :-P

-----


I'm totally OK with precircumfix `[]`, as I championed at the beginning.

Is there actually a case where we want it inside a type signature? `(a :: [B](...)) ->`? not sure offhand, but I'd say no.

**Comment by rhendric** [comment]
Why shouldn't it be valid there?

**Comment by rhendric** [comment]
In the precircumfix `[]` world, disambiguating between these cases would be tricky, I think:
* `o[x](a) -> b`
  * tokens: `ID:o DOT:. [ ID: x ] CALL(:( ID:a )CALL:) CALL(: PARAM(: )PARAM: -> ID:B )CALL:`
  * compiles to: `o[x](a)(function(){ return b; });`
* `o [x](a) -> b`
  * tokens: `ID:o CALL(: GENERIC_FUN: [ ID: x ] PARAM(:( ID:a )PARAM:) -> ID:B )CALL:`
  * compiles to: `o(function<x>(a) { return b; });`
* `o [x](a)`
  * tokens: `ID:o CALL(: [ ID: x ] CALL(:( ID:a )CALL:) )CALL:`
  * compiler error (or, would compile to: `o([x](a));`)

The extra `PARAM(: )PARAM:` pair gets inserted early in the lexer, so either a post-tokenize rewriter has to deal with that (in addition to moving one `CALL(:` and rewriting existing `CALL`s to `PARAM`s), or the initial tokenization has to be made more complicated to keep track of where the `[]` starts and whether there's a space before that (keeping in mind cases like `[[T](a :: T) -> ...]`).

**Comment by vendethiel** [comment]
> Why shouldn't it be valid there?

I wasn't sure ` fn<A>(a: <B>(x: A) => B): <B>(x: A) => B` was actual working syntax. Nice!

-----

> `o[x](a) -> b`
> `o [x](a)`

For both... We only need to look a single token (`]`) back for type parameters after we're done tagging `PARAM(` and `)PARAM` in `Lexer::parameters`, no? In this case, we'd look back, see `CALL(:` and stop there.
Is there an order issue I'm not taking into account?

**Comment by rhendric** [comment]
The point is that whether or not those parentheses should become `PARAM`s or `CALL`s depends on whether there's a space before the opening `[`. With the lexer as it is, they're assumed to be `CALL`s since they start adjacent to `]`, and `Lexer::parameters` will ignore them and introduce new `PARAM( )PARAM` tokens. But suppressing interpreting `(` as `CALL(` wouldn't work because it depends on whether there's going to be a `->` later. So `Lexer::parameters` has to look all the way back to `[`, or we need to track some fiddly state about the last `[` seen. Or fix the whole thing in a fiddly rewriter.

There could very well be a simpler answer—don't let me stop you looking for one. I just tried and that's what I ran into before I gave up for now.

**Comment by vendethiel** [comment]
Coming back to this (I'd like to have it for 1.7, if that's imaginable...)

regarding `[x,y]() ->`, it's currently an "invalid callee" error, so even if it's tough to fix in lexer (rewriter could probably go back to `[`, find if the previous is a `CALL(`, that means it's spaced... Though it also needs to handle `myfun = []() ->`)
 - maybe with a grammar production? `TypeList CALL( ... )`
 - in the nodes, if we have a Chain with a Arr followed by a Call (but we need to rewrite this to a signature...).

From what we discussed earlier, it seems we don't need a special syntax for `infer` which can be littered wherever the type variable is used.

**Comment by rhendric** [comment]
By all means, if you have an approach you think will work, give it a shot. I don't think I have any more helpful thoughts to offer you about the `[]() ->` problem—I'm a little pessimistic about the possibility of wedging that syntax into the parser we have, so for now I'm more excited about spending my free hacking time working on an entirely new parser, which may form the basis of a successor language. But I will still gladly review and offer feedback on patches if you write them.

**Comment by vendethiel** [comment]
OK. I tried for a bit, and the AST is a bit too damaged by the time we get to `Chain` in the AST. (the pattern is `Chain [Arr with generics; Call with the params; Call with the actual function definition]`.
Restoring it is definitely possible, but it's ugly.
I have a patch that implements `[] ::>` as the option you had suggested, and one that implements value-level `forall`.

```livescript
$ lscc 'f [A, B] ::> (a :: A, b :: B) ->'
f(function<A, B>(a: A, b: B): any{});
$ lsc 'f = [A, B] ::> (a :: A, b :: B) ->'
var f: <A, B>(arg0: A, arg1: B) => any;
f = function(a, b){};
```

I can take a stab at the `Chain` rewriting.

**Comment by rhendric** [comment]
Nice. Reviewing the thread, it occurs to me that my earlier discomfort with value-level `forall` also applies to my own proposal of `::>`, so the only real difference between the two is readability and the consumption of a new keyword. If we do run with `::>`, is there a reason to use square brackets there?

(Still kind of hoping one of us—or someone else—comes up with either an elegant implementation of `[]() ->` (somehow...) or a new syntax proposal that doesn't suck.)

**Comment by Djeg** [comment]
Hello :). I'm late and probably out of subject but anyway. What about a type system inspired by the [fantasy land](https://github.com/fantasyland/fantasy-land#type-signature-notation) specification and a bit of purescript ?

It could look like these:

```livescript
# var name: string = 'john'
let name :: String = 'John'

# var add: (a: number, b: number) => number;
# add = function (a, b) { return a + b; }
let add :: Number -> Number -> Number = (a, b) --> a + b

# multiline support and without curry
let sub
  :: (Number, Number) -> Number
  = (a, b) -> a - b

# Generic support:
let map
  :: a, b. (a -> b) -> Array a -> Array b
  = (f, arr) --> ...
# var map: <A, B>(f: ((a1: A) => B), arr: Array<A>) => Array<B>;
# map = function (f, arr) => ...

# Interfaced generic support:
let map
  :: User a, b. (a -> b) -> Array a -> Array b
  = (f, arr) --> ...
# var map: <A extends User, B>(f: ((a1: A) => B), arr: Array<A>) => Array<B>;
# map = function (f, arr) => ...
```

Wich i think open the door to some inteligent pattern matching (i'm going too far ^^):

```livescript
let addFiveToTen :: Number -> Number =
  | (10) -> 15
  | (a) -> a
```

By the way, very good work with livescript, i'm really enjoying it :)

**Comment by vendethiel** [comment]
What about the lambdas?

**Comment by Djeg** [comment]
Good point, like `do -> 3 + 4` ?
Maybe something like `do :: () -> Number = -> 2 + 4` ?

I'm not sure to get a concrete use case of a lambda here, do you have any concrete example ?

**Comment by vendethiel** [comment]
The whole thread goes into detail about our ideas. I know it’s fairly long but have you read all proposals?
